### PR TITLE
[omni, recipe, tests] feat: MiniCPM-o 4.5 thinker GSPO on the omni separate-async trainer

### PR DESCRIPTION
**Pull request by zhtmike**
# [omni, recipe, tests] feat: MiniCPM-o 4.5 thinker GSPO on the omni separate-async trainer

Implements #628. Branch: `feat/minicpmo-avqa-gspo-separate-async`, based on PR #572's
branch (`feat/minicpmo-avqa-gspo`).

### What does this PR do?

PR #572 shipped the MiniCPM-o 4.5 thinker-only AVQA GSPO recipe on the **collocated**
`omni_sync` trainer: one GPU pool hosts both the FSDP2 actor and the vLLM-Omni rollout,
generation and training strictly alternate, and merged LoRA weights re-travel to the
engine every step. PR #380 then landed the **disaggregated** trainer
(`trainer.v1.trainer_mode=omni_separate_async`): a standalone rollout pool on dedicated
GPUs generates one batch ahead of training through a replay buffer, and weights sync
every `parameter_sync_step` inner steps — validated on Qwen3-Omni (MMK12).

This PR ports the MiniCPM-o recipe onto that trainer (issue #628, the deferred
disaggregated item of #406). It adds **no trainer modes, no registry keys, no files
under `verl_omni/pipelines/`, and no engine changes** — a recipe, docs, and CPU tests:

- New recipe
  `examples/gspo_trainer/minicpm/run_minicpmo_4_5_thinker_gspo_lora_avqa_separate_async_v1.sh`:
  single node, 4 × 80GB split as 2 FSDP trainer + 2 rollout GPUs (**two standalone
  TP=1 replicas**, with one TP=2 replica — the validated Qwen3-Omni shape — as the
  documented config-only fallback). Batch identity `128 = parameter_sync_step(8) ×
  ppo_mini_batch_size(16)` matches **both** parent recipes; staleness is pinned
  `trainer.v1.sampler.max_off_policy_threshold=8` (one sync cycle) so off-policyness
  readings stay interpretable; backend `nccl`, `gpu_memory_utilization=0.8` (dedicated
  rollout GPUs), GSPO/LoRA/data/reward knobs inherited verbatim from #572 — except the
  colocated `param_offload`/`optimizer_offload` flags, dropped because the trainer GPUs
  are dedicated (see the README deltas table).
- LoRA ships **merged** (`lora.merge=True`, the same semantics as the colocated
  recipe) so the first disaggregated run stays directly comparable with the
  #572 reference for accuracy validation. The performance flip — `merge=False`,
  adapter deltas via `add_lora` through #380's LoRA-aware checkpoint engine,
  ~100 MB vs ~19 GB per sync — is analyzed and name-pinned as a later config-only
  change (see Design).
- README subsection + `docs/algo/separate_async_omni.md` update; three CPU test
  surfaces (see Test).

### Checklist Before Starting

- [x] Search for similar PRs:
  - `gh pr list --repo verl-project/verl-omni --state open --search "628 in:body"` → no open PR references the issue
  - `gh pr list --repo verl-project/verl-omni --state open --search "minicpm separate async in:title,body is:open"` → empty
  - #380 added the trainer and validated it on Qwen3-Omni only; the MiniCPM port and
    its model-specific wiring are new here, not duplicated anywhere.
- [x] Title format `[omni, recipe, tests] feat: ...` — modules and type validated
  against `tests/special_sanity/check_pr_title.py`.

### Test

CPU (macOS, miniforge `verl-omni-py312`; real observed output):

```
$ TORCH_COMPILE_DISABLE=1 TORCHINDUCTOR_DISABLE=1 python3 -m pytest -s --asyncio-mode=auto \
    tests/utils/test_minicpmo_gspo_separate_async_launcher_on_cpu.py \
    tests/pipelines/test_minicpm_lora_sync_names_on_cpu.py \
    tests/workers/rollout/test_omni_rollout_recovery_on_cpu.py
=> 12 passed

$ TORCH_COMPILE_DISABLE=1 TORCHINDUCTOR_DISABLE=1 python3 -m pytest -s --asyncio-mode=auto \
    tests/trainer/omni/test_ray_omni_trainer_separate_async_on_cpu.py \
    tests/utils/test_minicpmo_gspo_launcher_on_cpu.py \
    tests/workers/test_checkpoint_engine_on_cpu.py
=> 15 passed

$ pre-commit run --files <the six changed files>
=> ruff / ruff format / mypy / generated-yaml / doc / license / device-api /
   DataProto / validate-structure / naming / compileall — all Passed
```

### API and Usage Example

No API change — one new launch script:

```bash
bash examples/gspo_trainer/minicpm/run_minicpmo_4_5_thinker_gspo_lora_avqa_separate_async_v1.sh

# Any field stays overridable through "$@" extras:
bash examples/gspo_trainer/minicpm/run_minicpmo_4_5_thinker_gspo_lora_avqa_separate_async_v1.sh \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2   # one TP=2 replica fallback
bash examples/gspo_trainer/minicpm/run_minicpmo_4_5_thinker_gspo_lora_avqa_separate_async_v1.sh \
    actor_rollout_ref.model.lora.merge=False                 # adapter-delta perf flip
```

### Design & Code Changes

**The one load-bearing question — LoRA adapter-delta key alignment — resolves
natively, with no remap; the recipe ships merged sync first anyway.** The
separate-async trainer's efficient LoRA path sends peft adapter tensors over NCCL
and binds them on replicas via `add_lora`; the colocated recipe instead sends
merged full weights. This PR ships `merge=True` — merged full weights through the
same NCCL manager, the identical weight-naming the colocated recipe already
validates against this rollout class — so the first disaggregated run is directly
comparable with the #572 reference for accuracy validation; flipping to adapter
deltas later (a config-only perf change, ~100 MB vs ~19 GB per sync) is gated on
the naming question, which resolves without any remap:

- Actor side: LoRA targets `q/k/v/o_proj` inside `MiniCPMO.llm` (a
  `Qwen3ForCausalLM`), and the actor path injects peft into the wrapper itself
  (`non_diffusers_model_base.add_adapter` → `inject_adapter_in_model`), so
  `get_peft_model_state_dict` emits `llm.model.layers.N.self_attn.q_proj.lora_A/B.weight`.
- Rollout side: `MiniCPMO45OmniLLMForConditionalGeneration` instantiates the same
  backbone under the `llm.` prefix (`maybe_prefix(prefix, "llm")` +
  `init_vllm_registered_model`, `vllm_omni/.../minicpmo_4_5_omni_llm.py`), and vLLM's
  `parse_fine_tuned_lora_name` resolves `...lora_A.weight` onto exactly those names.
- Nothing rewrites the keys in between: MiniCPM-o defines no
  `_checkpoint_conversion_mapping` (unlike Qwen3-Omni, whose thinker-prefix strip is
  what aligns *that* model), so verl's `convert_weight_keys` is an identity pass.

This is pinned on CPU by the new `tests/pipelines/test_minicpm_lora_sync_names_on_cpu.py`
(a tiny `MiniCPMO`-shaped wrapper + peft injection, resolved through vLLM's real
`parse_fine_tuned_lora_name` against the rollout-side naming; also proves the recipe's
`exclude_modules` regex keeps the frozen `vpm`/`apm` towers — which carry their own
`q_proj` modules — out of the adapter). The later `merge=False` flip is therefore
one config line, documented in the README.

**Changed files:**

- `examples/gspo_trainer/minicpm/run_minicpmo_4_5_thinker_gspo_lora_avqa_separate_async_v1.sh`
  (new): the Qwen3-Omni MMK12 separate-async script's structure (it encodes the three
  trainer-startup asserts: rollout GPUs > 0, non-naive backend, batch identity) carrying
  #572's AVQA model deltas verbatim.
- `examples/gspo_trainer/README.md`: "Run disaggregated (separate-async)" subsection —
  deltas table vs the colocated recipe, the TP fallback, the LoRA naming note, and the
  reward-placement rule under this trainer mode (future GPU reward models go on the
  standalone/own pool, never the trainer pool; this recipe's rule-based `minicpm_naive`
  reward needs no placement).
- `docs/algo/separate_async_omni.md`: MiniCPM recipe listed alongside Qwen3-Omni, the
  per-mode LoRA send divergence recorded as resolved, new test commands.
- `tests/utils/test_minicpmo_gspo_separate_async_launcher_on_cpu.py` (new): launcher
  contract — trainer mode, the `128 == 8 × 16` identity computed from the parsed
  settings, staleness threshold == sync step, GPU split, `backend=nccl`,
  `merge=True` (accuracy-parity choice, with the `merge=False` perf flip noted),
  `load_format=safetensors`, MiniCPM wiring, `VERL_USE_EXTERNAL_MODULES` export,
  and the negative guards
  (no `override_config`, no sdpa, no non-val sampling overrides) mirroring the
  colocated recipe's test.
- `tests/pipelines/test_minicpm_lora_sync_names_on_cpu.py` (new): the adapter-delta
  naming pin described above, plus `convert_weight_keys` identity for the wrapper.
- `tests/workers/rollout/test_omni_rollout_recovery_on_cpu.py`: adds an AVQA-shaped
  abort/continuation case (image + audio payloads and the 16 kHz sampling kwarg must
  ride every resubmission across a weight-sync abort, budget shrinking by the tokens
  already generated) alongside #380's existing generic case.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] `pre-commit run --files` on all six changed files — all hooks Passed (see Test).
- [x] Docs updated: examples README + `docs/algo/separate_async_omni.md`.
- [x] CPU unit tests added covering the recipe contract and the adapter-delta naming
  alignment; no GPU e2e is added to CI — the repo's CI is CPU-only, and MiniCPM-o is a
  `trust_remote_code` checkpoint with no tiny-random builder (same scope decision as
  PR #572, which validated on real hardware outside CI).

---

**AI assistance (ZCode) was used for this change.** A human submitter has reviewed
every changed line and ran the commands above.

### CODE REVIEW

**Comment by SamitHuang** [comment]
Is there any convergence test results?

**Comment by zhtmike** [comment]
> Is there any convergence test results?

uploaded.

**Comment by zhtmike** [comment]
**Perf study: colocated sync vs separated async — MiniCPM-o 4.5 AVQA, 4×80 GB GPUs**

Characterized the separate-async deployment against colocated training. Short version: **on this workload (short ~100-token responses) colocated wins; separated async shows no bugs and no pipeline bubbles, and flips to winning as generation grows longer / more long-tailed.**

**Setup**: identical recipe (`examples/gspo_trainer/minicpm/run_minicpmo_4_5_thinker_gspo_lora_avqa*.sh`), only deployment lines differ; 45-step timing probes at constant LR, matched response lengths (~90–100 tok).

| Arm | Layout | Step time | tok/s/GPU |
|---|---|---|---|
| colocated sync | 4-GPU FSDP + 2×TP2 rollout colocated (sleep/wake) | **181.5 s** | **~1650** |
| separate async 3+1 | 3-GPU FSDP + 1×TP1 standalone rollout | 192.2 s | 1543 |
| separate async 2+2 | 2-GPU FSDP + 2×TP1 standalone rollout | ~273 s | ~1150 |

Evidence: wandb [xed2e3ly](https://wandb.ai/verl-omni/gspo/runs/xed2e3ly) / [8xxsb4dr](https://wandb.ai/verl-omni/gspo/runs/8xxsb4dr) / [gq15fn5v](https://wandb.ai/verl-omni/gspo/runs/gq15fn5v) (plus original full runs [b06hw7nk](https://wandb.ai/verl-omni/gspo/runs/b06hw7nk) / [6o1rdea9](https://wandb.ai/verl-omni/gspo/runs/6o1rdea9)).

**Why colocated wins here**: the workload is trainer-bound (~2–3× more training than generation GPU-seconds per step). Colocated time-multiplexes all 4 GPUs into each phase; separation trades trainer FLOPs for overlap. At 3+1, overlap hides ~50 s of generation but the training chain inflates 130→192 s (4→3 FSDP GPUs) → ~6% slower. At 2+2 it's far worse: −33% trainer FLOPs → ~33% slower overall, with rollout GPUs 84% idle. GPU duty (dmon, steady state): colocated 76–77% on all 4, symmetric, ~19 s/step fully dark (sleep/wake + weight sync — exactly the overhead async removes); 3+1 → 70% trainer / 39% rollout; 2+2 → 77% / 16%.

Separation becomes the better choice when generation dominates (long/thinking/agentic rollouts) or when long generation tails stall the colocated fleet — colocated steps were observed stretching to 245 s on long-response batches, while async absorbed the same tails with ~0.4 s generation wait and staleness pinned at 1.

**3+1 configurability note**: verl requires `ppo_mini_batch_size × rollout.n` (sequences per inner update) divisible by the trainer DP size. The default 16×16 = 256 fails for DP=3; use `data.train_batch_size=120` + `actor_rollout_ref.actor.ppo_mini_batch_size=15` (240 seqs; 120 = 8×15 preserves the `parameter_sync_step` invariant and the 8-updates/step cadence).

**Health audit of the async path** (no perf bugs found):
- The abort→resume→continue flow around weight sync is correct — partial tokens are kept (`prompt + partial_tokens`), the token budget is re-clamped across the boundary, and continuation fully re-prefills under the new weights. It never fired in these runs (generation always finished before the sync; `trajectory_spans` = 1 throughout).
- Observability gaps (not correctness): `aborted_count` is discarded by the checkpoint engine, and per-sample rollout timers are dropped by the V1 agent-loop manager — abort/regeneration is invisible in wandb today.
- Multi-stage (non-AR) abort is broken upstream — talker/multi-stage output modes should not use separate async yet.

@SamitHuang @knlnguyen1802

**Comment by zhtmike** [comment]
Metric definitions (verl `timing_s/*` wandb keys):

- **step time** (`timing_s/step`) — wall-clock of one full outer training step: generate one batch of prompts × 16 responses, run all policy updates, sync weights back to the rollout engine.
- **gen** (`timing_s/gen`) — time the trainer *waits* for rollout samples. In colocated sync this equals the whole generation phase (it is serialized with training); in separated async generation runs concurrently with training, so ~0 means fully hidden.
- **old_log_prob** (`timing_s/old_log_prob`) — recomputing log-probs of the sampled responses under the actor policy (the π_old anchor for GSPO/PPO ratios). In async mode this also includes CPU save/restore of π_old between inner updates (decoupled PPO).
- **adv** (`timing_s/adv`) — advantage computation (GRPO group normalization).
- **update_actor** (`timing_s/update_actor`) — the policy-gradient optimizer updates (8 per outer step in all arms).
- **update_weights** (`timing_s/update_weights`) — pushing updated weights to the rollout engine: in colocated, the in-process sleep/wake handover; in separated async, the abort → KV-release → NCCL broadcast → resume cycle.
- **trainer chain** — old_log_prob + adv + update_actor + update_weights (the trainer-side critical path).
- **throughput** (`perf/throughput`) — prompt+response tokens per second per GPU, divided by all 4 GPUs in every arm (directly comparable).
- **GPU duty** — fraction of steady-state wall time with SM utilization > 15% (`nvidia-smi dmon`, 1 s sampling), per role.
- **staleness** (`training/off_policy/trajectory_spans` max) — how many weight versions a trajectory spanned; 1 = generated entirely under one policy version (i.e., on-policy; the abort-and-continue mechanism never had to fire).

**Detailed measurements (steady-state means, seconds unless noted)**

| Metric | colocated 4+0 (120×16) | async 3+1 (120×16) | async 2+2 (128×16) | colocated 4+0 full run (128×16) | async 2+2 full run (128×16) |
|---|---|---|---|---|---|
| step time | **181.5** | 192.2 | ~273 | 207.5 | 283.1 |
| gen: serial phase / trainer wait | 50.5 serial | 0.37 wait | 0.39 wait | 69.1 serial | 0.4 wait |
| old_log_prob | 43.0 | 88.9 | ~124 | 46.2 | 131.9 |
| adv | 1.3 | 1.0 | 1.1 | 1.7 | 1.1 |
| update_actor | 77.5 | 92.6 | ~137 | 81.3 | 140.1 |
| update_weights | 8.7 | 9.3 | ~8.7 | 8.5 | 8.7 |
| trainer chain total | 130.5 | 191.8 | ~271 | 137.7 | 281.8 |
| throughput (tok/s/GPU) | **~1650** | 1543 | ~1150 | 1562 | 1129 |
| GPU duty: trainer / rollout (dmon) | 76–77% all 4, symmetric | 70% / 39% | 77% / 16% | — | — |
| staleness (`trajectory_spans` max) | on-policy | 1 | 1 | on-policy | 1 |
| wandb run | [xed2e3ly](https://wandb.ai/verl-omni/gspo/runs/xed2e3ly) | [8xxsb4dr](https://wandb.ai/verl-omni/gspo/runs/8xxsb4dr) | [gq15fn5v](https://wandb.ai/verl-omni/gspo/runs/gq15fn5v) | [b06hw7nk](https://wandb.ai/verl-omni/gspo/runs/b06hw7nk) | [6o1rdea9](https://wandb.ai/verl-omni/gspo/runs/6o1rdea9) |

Measurement windows: colocated 120 excludes 6 gen-straggler steps (73–108 s gen; incl. them: 188.6 s mean); async 3+1 = steps 10–45; async 2+2 probe = steps 2–24 of a 45-step probe; full runs = whole-history steady-state means (response lengths drift upward late in training, which inflates them ~5%). Batch notation: `train_batch_size × rollout.n`. Trainer chain = old_log_prob + adv + update_actor + update_weights.
